File: frontend/src/api/images.ts
import apiClient from './client'

export async function uploadImage(file: File, pageId: string) {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await apiClient.post(`/images/upload?page_id=${pageId}`, formData)
  return data as { image_key: string; url: string }
}

File: backend/app/api/routes/images.py
from fastapi import APIRouter, File, Query, UploadFile

from app.services.storage import upload_image, InvalidFileError

router = APIRouter(prefix="/images", tags=["images"])


@router.post("/upload")
async def upload_image_endpoint(
    file: UploadFile = File(...),
    page_id: str = Query(..., description="Page ID for key namespacing"),
):
    """Upload an image to MinIO storage."""
    content = await file.read()
    try:
        result = await upload_image(
            file_content=content,
            filename=file.filename or "image.bin",
            content_type=file.content_type or "",
            page_id=page_id,
        )
    except InvalidFileError as e:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail=str(e))
    return result
